File: 112. Path Sum/hasPathSum.py
from typing import Optional
# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

class Solution:
  def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
    def dfs(node, currentSum):
      if not node:
        return False
      
      currentSum += node.val
      if not node.left and not node.right:
        return currentSum == targetSum
      
      return dfs(node.left, currentSum) or dfs(node.right, currentSum)
    
    return dfs(root, 0)
  

root = TreeNode(5)
root.left = TreeNode(4)
root.right = TreeNode(8)

root.left.left = TreeNode(11)

root.left.left.left = TreeNode(7)
root.left.left.right = TreeNode(2)

root.right.left = TreeNode(13)
root.right.right = TreeNode(4)

root.right.right.right = TreeNode(1)

solution = Solution()
print('hasPathSum: ', solution.hasPathSum(root, 22))

File: 112. Path Sum/hasPathSum.js
// Definition for a binary tree node.
class TreeNode {
  constructor(val = 0, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */
const hasPathSum = (root, targetSum) => {
  const dfs = (node, currentSum) => {
    if (!node) {
      return false;
    }

    currentSum += node.val;
    if (!node.left && !node.right) {
      return currentSum === targetSum;
    }

    return dfs(node.left, currentSum) || dfs(node.right, currentSum);
  }

  return  dfs(root, 0);
};

const root = new TreeNode(5);
root.left = new TreeNode(4);
root.right = new TreeNode(8);

root.left.left = new TreeNode(11);

root.left.left.left = new TreeNode(7);
root.left.left.right = new TreeNode(2);

root.right.left = new TreeNode(13);
root.right.right = new TreeNode(4);

root.right.right.right = new TreeNode(1);

console.log('hasPathSum: ', hasPathSum(root, 22));
